lib/AST/ASTScopeCreation.cpp

```diff
@@ -0,0 +1,1061 @@
+  /// was originally built, add them as children of this scope.
+  NullablePtr<ASTScopeImpl> lastAdopter;
+
+  /// Information about how this object was created for debugging.
+  const char *const file;
+  const int line;
+  const ASTScopeImpl &creator;
+  const void *const forWhat;
+
+public:
+  DeferredNodes(const char *file, int line, const ASTScopeImpl &creator,
+                const void *forWhat);
+  ~DeferredNodes();
+  DeferredNodes(const DeferredNodes &) = delete;  // ensure no copies
+  DeferredNodes(const DeferredNodes &&) = delete; // ensure no moves
+
+  bool empty() const;
+
+  void pushIfNecessary(ASTNode);
+  template <typename ASTNodelike>
+  void pushAllNecessaryNodes(ArrayRef<ASTNodelike>);
+  void pushSourceFileDecls(ArrayRef<Decl *>);
+
+  ASTNode popNextDeferredNodeForAdoptionBy(ASTScopeImpl *);
+
+  NullablePtr<ASTScopeImpl> getLastAdopter() const;
+  void setLastAdopter(ASTScopeImpl *);
+
+  void dump() const;
+  void print(raw_ostream &) const;
+
+  // Make vanilla new illegal for ASTScopes.
+  void *operator new(size_t bytes) = delete;
+  // Need this because have virtual destructors
+  void operator delete(void *data) {}
+
+  // Only allow allocation of scopes using the allocator of a particular source
+  // file.
+  void *operator new(size_t bytes, const ASTContext &ctx,
+                     unsigned alignment = alignof(DeferredNodes));
+  void *operator new(size_t Bytes, void *Mem) {
+    assert(Mem);
+    return Mem;
+  }
+};
+} // ast_scope
+} // namespace swift
+using namespace ast_scope;
+
+#pragma mark Scope tree creation and extension
+
+ASTSourceFileScope::ASTSourceFileScope(SourceFile *SF)
+    : SF(SF), deferred(new (SF->getASTContext())
+                           DeferredNodes(__FILE__, __LINE__, *this, SF)) {}
+
+ASTSourceFileScope *ASTSourceFileScope::createScopeTreeFor(SourceFile *SF) {
+  auto *sfScope = new (SF->getASTContext()) ASTSourceFileScope(SF);
+  sfScope->addAnyNewDeclsToTree();
+  return sfScope;
+}
+
+void ASTSourceFileScope::addAnyNewDeclsToTree() {
+  deferred->pushSourceFileDecls(SF->Decls);
+  auto lastAdopter = deferred->getLastAdopter();
+  auto *adopter = lastAdopter ? lastAdopter.get() : this;
+  adopter->addChildScopesForAnyRemainingDeferredNodes(*deferred);
+}
+
+#pragma mark ScopeCreator
+
+namespace swift {
+namespace ast_scope {
+
+class ScopeCreator
+    : public ASTVisitor<ScopeCreator, void, void, void, void, void, void,
+                        ASTScopeImpl *, DeferredNodes &> {
+public:
+
+#pragma mark ASTNodes that do not create scopes
+
+  // Even ignored Decls and Stmts must extend the source range of a scope:
+  // E.g. a braceStmt with some definitions that ends in a statement that
+  // accesses such a definition must resolve as being IN the scope.
+  // clang-format off
+                          
+  void visitImportDecl(                       ImportDecl *d, ASTScopeImpl *p, DeferredNodes&) { p->widenSourceRangeForIgnoredASTNode(d); }
```

Comment:
Can you use a `#define` here to avoid reproducing the code? E.g.:

```
#ifdef SCOPE_CREATOR_IGNORED
#error "Already created"
#endif
#define SCOPE_CREATOR_IGNORED(NAME) \
void visit##Name(Name *d, ASTScopeImpl *p, DeferredNodes &) { \
  p->widenSourceRangeForIgnoredASTNode(d); \
}
SCOPE_CREATOR_IGNORED(ImportDecl)
SCOPE_CREATOR_IGNORED(EnumCaseDecl)
...
#undef SCOPE_CREATOR_IGNORED
```

Reply:
Yes, I had wondered about how folks here feel about defines? Makes the code shorter but harder to grep.

Reply:
OK, added defines (not pushed yet)